src/app/clusters/scenes-server/SceneHandlerImpl.cpp

```diff
@@ -16,10 +16,224 @@
+/// @brief Helper function to convert a byte array to a value of the given type.
+/// @param EmberAfDefaultAttributeValue & defaultValue
+/// @return Value converted to the given working type
+template <typename Type>
+typename app::NumericAttributeTraits<Type>::WorkingType
+ConvertDefaultValueToWorkingValue(const EmberAfDefaultAttributeValue & defaultValue)
+{
+    if (sizeof(typename app::NumericAttributeTraits<Type>::WorkingType) <= 2)
+    {
+        return static_cast<typename app::NumericAttributeTraits<Type>::WorkingType>(defaultValue.defaultValue);
+    }
+
+    typename app::NumericAttributeTraits<Type>::StorageType sValue;
+    memcpy(&sValue, defaultValue.ptrToDefaultValue, sizeof(sValue));
+    return app::NumericAttributeTraits<Type>::StorageToWorking(sValue);
+}
+
+/// IsOnlyOneValuePopulated
+/// @brief Helper function to verify if only one value is populated in a given AttributeValuePairType
+/// @param AttributeValuePairType & type AttributeValuePairType to verify
+/// @return bool true if only one value is populated, false otherwise
+bool IsOnlyOneValuePopulated(const AttributeValuePairType & type)
+{
+    int count = 0;
+    if (type.valueUnsigned8.HasValue())
+        count++;
+    if (type.valueSigned8.HasValue())
+        count++;
+    if (type.valueUnsigned16.HasValue())
+        count++;
+    if (type.valueSigned16.HasValue())
+        count++;
+    if (type.valueUnsigned32.HasValue())
+        count++;
+    if (type.valueSigned32.HasValue())
+        count++;
+    if (type.valueUnsigned64.HasValue())
+        count++;
+    if (type.valueSigned64.HasValue())
+        count++;
+    return count == 1;
+}
+
+/// CapAttributeID
+/// Cap the attribute value based on the attribute's min and max if they are defined,
+/// or based on the attribute's size if they are not.
+/// @param[in] aVPair   AttributeValuePairType
+/// @param[in] metadata  EmberAfAttributeMetadata
+///
+template <typename Type>
+void CapAttributeID(typename app::NumericAttributeTraits<Type>::WorkingType & Value, const EmberAfAttributeMetadata * metadata)
+{
+    using IntType     = app::NumericAttributeTraits<Type>;
+    using WorkingType = typename IntType::WorkingType;
+
+    WorkingType maxValue;
+    WorkingType minValue;
+    uint16_t bitWidth = static_cast<uint16_t>(emberAfAttributeSize(metadata) * 8);
+
+    // Min/Max Value capps for the OddSize integers
+    if (metadata->IsSignedIntegerAttribute())
+    {
+        // We use emberAfAttributeSize for cases like INT24S, INT40S, INT48S, INT56S where numeric_limits<WorkingType>::max()
+        // wouldn't work
+        maxValue = static_cast<WorkingType>((1ULL << (bitWidth - 1)) - 1);
+        minValue = static_cast<WorkingType>(-(1ULL << (bitWidth - 1)));
+    }
+    else
+    {
+        // We use emberAfAttributeSize for cases like INT24U, INT40U, INT48U, INT56U where numeric_limits<WorkingType>::max()
+        // wouldn't work
+        if (ZCL_INT64U_ATTRIBUTE_TYPE == metadata->attributeType)
```

Comment:
This is going to fail badly for attributes with types like ZCL_BITMAP64_ATTRIBUTE_TYPE, ZCL_NODE_ID_ATTRIBUTE_TYPE, etc, etc.

Followup: we need to not be doing any of this stuff with metadata.  We already know the type: it's Type.  The min/max should just come from that.

Reply:
Added check on AttributeBaseType for now.